drivers/power_domain/power_domain_gpio_monitor.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT power_domain_gpio_monitor
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(power_domain_gpio_monitor, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+struct pd_gpio_monitor_config {
+	struct gpio_dt_spec power_good_gpio;
+};
+
+struct pd_gpio_monitor_data {
+	struct gpio_callback callback;
+	bool is_powered;
+	const struct device *dev;
+};
+
+struct pd_visitor_context {
+	const struct device *domain;
+	enum pm_device_action action;
+};
+
+static int pd_on_domain_visitor(const struct device *dev, void *context)
+{
+	struct pd_visitor_context *visitor_context = context;
+
+	/* Only run action if the device is on the specified domain */
+	if (!dev->pm || (dev->pm->domain != visitor_context->domain)) {
+		return 0;
+	}
+
+	dev->pm->usage = 0;
+	(void)pm_device_action_run(dev, visitor_context->action);
+	return 0;
+}
+
+static int pd_gpio_monitor_update_status(struct pd_gpio_monitor_data *data)
+{
+	struct device *dev = data->dev;
+	struct pd_visitor_context context = {.domain = dev};
+	const struct pd_gpio_monitor_config *config = data->dev->config;
+	int rc;
+
+	rc = gpio_pin_get_dt(&config->power_good_gpio);
+	if (rc < 0) {
+		LOG_WRN("Failed to read gpio logic level");
+		return rc;
+	}
+
+	data->is_powered = !!rc;
```

Comment:
Don't thnk you need the `!!`, `is_powered` is a boolean, the compiler should do the right thing automatically.

Reply:
Done